File: widgetssdk/src/test/java/com/glia/widgets/filepreview/domain/GetImageFileFromNetworkUseCaseTest.java
package com.glia.widgets.filepreview.domain;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import android.graphics.Bitmap;

import com.glia.androidsdk.chat.AttachmentFile;
import com.glia.widgets.chat.domain.DecodeSampledBitmapFromInputStreamUseCase;
import com.glia.widgets.filepreview.data.GliaFileRepository;
import com.glia.widgets.filepreview.domain.exception.FileNameMissingException;
import com.glia.widgets.filepreview.domain.exception.RemoteFileIsDeletedException;
import com.glia.widgets.filepreview.domain.usecase.GetImageFileFromNetworkUseCase;

import org.junit.Before;
import org.junit.Test;

import java.io.IOException;
import java.io.InputStream;

import io.reactivex.Maybe;

public class GetImageFileFromNetworkUseCaseTest {

    private GliaFileRepository fileRepository;
    private AttachmentFile attachmentFile;
    private GetImageFileFromNetworkUseCase useCase;

    private static final InputStream INPUT_STREAM = mock(InputStream.class);
    private static final String NAME = "NAME";
    private static final String NAME_EMPTY = "";
    private static final Bitmap BITMAP = mock(Bitmap.class);
    private DecodeSampledBitmapFromInputStreamUseCase decodeSampledBitmapFromInputStreamUseCase;

    @Before
    public void setUp() {
        fileRepository = mock(GliaFileRepository.class);
        attachmentFile = mock(AttachmentFile.class);
        decodeSampledBitmapFromInputStreamUseCase = mock(DecodeSampledBitmapFromInputStreamUseCase.class);
        useCase = new GetImageFileFromNetworkUseCase(fileRepository, decodeSampledBitmapFromInputStreamUseCase);
    }

    @Test
    public void execute_emitsFileNameMissingException_whenFileIsNull() {
        useCase.invoke(null)
                .test()
                .assertError(FileNameMissingException.class);
    }

    @Test
    public void execute_emitsFileNameMissingException_whenFileNameIsEmpty() {
        when(attachmentFile.getName()).thenReturn(NAME_EMPTY);
        useCase.invoke(attachmentFile)
                .test()
                .assertError(FileNameMissingException.class);
    }

    @Test
    public void execute_emitsError_whenDecodingFails() {
        when(fileRepository.loadImageFileFromNetwork(any())).thenReturn(Maybe.just(INPUT_STREAM));
        when(attachmentFile.getName()).thenReturn(NAME);
        when(attachmentFile.isDeleted()).thenReturn(false);
        when(decodeSampledBitmapFromInputStreamUseCase.invoke(INPUT_STREAM))
                .thenReturn(Maybe.error(new IOException()));

        useCase.invoke(attachmentFile)
                .test()
                .assertError(IOException.class);
    }

    @Test
    public void execute_emitsRemoteFileIsDeletedException_whenFileIsDeleted() {
        when(attachmentFile.getName()).thenReturn(NAME);
        when(attachmentFile.isDeleted()).thenReturn(true);
        useCase.invoke(attachmentFile)
                .test()
                .assertError(RemoteFileIsDeletedException.class);
    }

    @Test
    public void execute_successfullyCompletes_whenValidArgument() {
        when(fileRepository.loadImageFileFromNetwork(any())).thenReturn(Maybe.just(INPUT_STREAM));
        when(decodeSampledBitmapFromInputStreamUseCase.invoke(INPUT_STREAM)).thenReturn(Maybe.just(BITMAP));
        when(attachmentFile.getName()).thenReturn(NAME);
        when(attachmentFile.isDeleted()).thenReturn(false);
        useCase.invoke(attachmentFile).test().assertResult(BITMAP);
    }
}

File: widgetssdk/src/test/java/com/glia/widgets/filepreview/domain/GetImageFileFromNetworkUseCaseTest.kt
package com.glia.widgets.filepreview.domain

import android.graphics.Bitmap
import com.glia.androidsdk.chat.AttachmentFile
import com.glia.widgets.chat.domain.DecodeSampledBitmapFromInputStreamUseCase
import com.glia.widgets.filepreview.data.GliaFileRepository
import com.glia.widgets.filepreview.domain.exception.FileNameMissingException
import com.glia.widgets.filepreview.domain.exception.RemoteFileIsDeletedException
import com.glia.widgets.filepreview.domain.usecase.GetImageFileFromNetworkUseCase
import io.reactivex.rxjava3.core.Maybe
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import java.io.IOException
import java.io.InputStream

class GetImageFileFromNetworkUseCaseTest {
    private lateinit var fileRepository: GliaFileRepository
    private lateinit var attachmentFile: AttachmentFile
    private lateinit var useCase: GetImageFileFromNetworkUseCase
    private lateinit var decodeSampledBitmapFromInputStreamUseCase: DecodeSampledBitmapFromInputStreamUseCase

    @Before
    fun setUp() {
        fileRepository = mock()
        attachmentFile = mock()
        decodeSampledBitmapFromInputStreamUseCase = mock()
        useCase = GetImageFileFromNetworkUseCase(
            fileRepository,
            decodeSampledBitmapFromInputStreamUseCase
        )
    }

    @Test
    fun execute_emitsFileNameMissingException_whenFileIsNull() {
        useCase(null)
            .test()
            .assertError(FileNameMissingException::class.java)
    }

    @Test
    fun execute_emitsFileNameMissingException_whenFileNameIsEmpty() {
        whenever(attachmentFile.name) doReturn NAME_EMPTY
        useCase(attachmentFile)
            .test()
            .assertError(FileNameMissingException::class.java)
    }

    @Test
    fun execute_emitsError_whenDecodingFails() {
        whenever(fileRepository.loadImageFileFromNetwork(any())) doReturn Maybe.just(INPUT_STREAM)
        whenever(attachmentFile.name) doReturn NAME
        whenever(attachmentFile.isDeleted) doReturn false
        whenever(decodeSampledBitmapFromInputStreamUseCase(INPUT_STREAM)) doReturn Maybe.error(
            IOException()
        )
        useCase(attachmentFile)
            .test()
            .assertError(IOException::class.java)
    }

    @Test
    fun execute_emitsRemoteFileIsDeletedException_whenFileIsDeleted() {
        whenever(attachmentFile.name) doReturn NAME
        whenever(attachmentFile.isDeleted) doReturn true
        useCase(attachmentFile)
            .test()
            .assertError(RemoteFileIsDeletedException::class.java)
    }

    @Test
    fun execute_successfullyCompletes_whenValidArgument() {
        whenever(fileRepository.loadImageFileFromNetwork(any())) doReturn Maybe.just(INPUT_STREAM)
        whenever(decodeSampledBitmapFromInputStreamUseCase(INPUT_STREAM)) doReturn Maybe.just(BITMAP)
        whenever(attachmentFile.name) doReturn NAME
        whenever(attachmentFile.isDeleted) doReturn false
        useCase(attachmentFile).test().assertResult(BITMAP)
    }

    companion object {
        private val INPUT_STREAM: InputStream = mock()
        private const val NAME = "NAME"
        private const val NAME_EMPTY = ""
        private val BITMAP: Bitmap = mock()
    }
}
